### PR TITLE
Propose Changes to Cloud Provider KEPs

### PR DESCRIPTION
**Pull request by andrewsykim**
* removes any duplicate information in KEP-0002 from KEP-00013
* reduces scope in KEP-00013 to only include proposal for removing in-tree provider code. Renaming the file removed the diff, see last commit to see diff between files.

### CODE REVIEW

File: keps/sig-cloud-provider/0002-cloud-controller-manager.md [hunk 1]
```diff
@@ -79,11 +79,8 @@ Controlled by [publishing kubernetes-rules-configmap](https://github.com/kuberne
 
 ## Summary
 
-We want to remove any cloud provider specific logic from the kubernetes/kubernetes repo. We want to restructure the code
-to make it easy for any cloud provider to extend the kubernetes core in a consistent manner for their cloud. New cloud
-providers should look at the [Creating a Custom Cluster from Scratch](https://kubernetes.io/docs/getting-started-guides/scratch/#cloud-provider)
-and the [cloud provider interface](https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/cloud.go#L31)
-which will need to be implemented.
+This KEP outlines the architectural changes required to run the Cloud Controller Manager. This is part of a long-running effort
```
Comment (cheftako) [inline keps/sig-cloud-provider/0002-cloud-controller-manager.md]: At some point I believe we need to expand the scope. CCM is a mechanism for allowing cloud specific controllers to be run in an K8s approved manner. I feel that our goal though is to provide a mechanism to allow every cloud provider (and in fact non cloud providers as well) to create installable Kubernetes which run correctly on their configurations. We would like to do that without including any cloud provider code in the main repo. To that end the CCM is necessary but not sufficient and we need to start tracking the additional pieces.

**Comment by hogepodge** [comment]
This PR is over a month old, but I have no objections to it. Where do we stand on it?

/lgtm

File: keps/sig-cloud-provider/0013-removing-in-tree-providers.md [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+The biggest challenge of this phase is to remove dependences to kubernetes/kubernetes in all the providers. This is required to avoid circular dependencies within the various Kubernetes repos. All other repos "staged" (`client-go`, `apimachinery`, `api`, etc) in Kubernetes follow the same pattern. Below are the current list of packges in `kubernetes/kubernetes` that we need to removed. Note that _some_ dependencies may need to be duplicated for the duration of this migration as other critical components of Kubernetes may share the same code but removing the dependency may not be trivial. In general, we will avoid duplicating code and only use it as a last resort to resolve circular dependencies to `kubernetes/kubernetes`.
+
+| Dependency                        | Proposed Location            | Is duplicated? | Partial or Full? |
```
Comment (andrewsykim) [inline keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: Need help filling this out, if you are assigned to any of the dependencies in https://github.com/kubernetes/kubernetes/issues/69585 please comment here on what the new locations should be :) thanks!

cc @kubernetes/sig-cloud-provider @cheftako @dims @d-nishi @frapposelli @hogepodge @islinwb @mcrute @stewart-yu @yastij

**Comment by andrewsykim** [comment]
Updated the KEP to better represent the approach we are taking now. Hoping it outlines our intentions & steps clearly! Some of the dependencies we want to move don't have a new home yet but hoping we can come to consensus for those here :). Please take a look! cc @cheftako @d-nishi @hogepodge @jagosan @thockin

File: keps/sig-cloud-provider/0013-removing-in-tree-providers.md (hunk 2 above)
Comment (andrewsykim) [reply keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: Realizing now that the `Partial vs Full` part is confusing, but this just indicating whether we need to migrate the entire package for our purposes or if we can just pick out the specific methods/types that we need.

**Comment by andrewsykim** [comment]
Thanks Tim! 💙

File: keps/sig-cloud-provider/0013-removing-in-tree-providers.md [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+2. Begin to build CCM from external repos (`kubernetes/cloud-provider-<provider>`) via code synced from `kubernetes/kubernetes/staging/cloud-provider-<provider>`. This allows us to develop providers "out-of-tree" while being able to build both in-tree components like kube-controller-manager and out-of-tree components like cloud-controller-manager. Development is still done in `kubernetes/kubernetes/staging/cloud-provider-<provider>`.
+3. Delete all code in `kubernetes/kubernetes/staging/cloud-provider-<provider>` and shift main development to `kubernetes/cloud-provider-<provider>`.
+
+#### Phase 1 - Moving Cloud Provider Code to Staging
+
+In Phase 1, all cloud provider code in `kubernetes/kubernetes/pkg/cloudprovider/providers` will be moved to `kubernetes/kubernetes/staging/cloudprovider/providers`. Moving code to a staging directory still allows us to build core Kubernetes components (`kube-controller-manager`, `kubelet`, etc) without interruption while also signalling to the community that in-tree provider code is slated for removal.
+
```
Comment (stewart-yu) [inline keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: `k8s.io/kubernetes/staging/cloud-provider/providers`?

File: keps/sig-cloud-provider/0013-removing-in-tree-providers.md (hunk 2 above)
Comment (stewart-yu) [reply keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: `pkg/controller` need more investgate, some api need move from `k8s.io/kubernetes/pkg/apis/core`, and code from "k8s.io/kubernetes/pkg/apis/core/helper" . i'll update later

File: keps/sig-cloud-provider/0013-removing-in-tree-providers.md (hunk 3 above)
Comment (andrewsykim) [reply keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: thanks

File: keps/sig-cloud-provider/0013-removing-in-tree-providers.md (hunk 2 above)
Comment (andrewsykim) [reply keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: For `pkg/controller`, it seems like we already copied [ControllerClientBuilder](https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L31-L40) and the method `AddOrUpdateTaintOnNode` can be moved to `k8s.io/cloud-provider/nodes`. What do you think?

Comment (andrewsykim) [reply keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: ah, sorry I forgot about dependency to `k8s.io/kubernetes/pkg/util/taints`

Comment (andrewsykim) [reply keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: Seems like `AddOrUpdateTaint` is only used in `pkg/controller` so might be reasonable to just move that method into `k8s.io/cloud-provider/nodes` as well.

Comment (stewart-yu) [reply keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: i move part of  `taint` op from  `k8s.io/kubernetes/pkg/util/taints` to `k8s.io/apimachinery/pkg/util/taints`. see 
https://github.com/kubernetes/kubernetes/pull/70225

Comment (stewart-yu) [reply keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: so form table as below:
`pkg/controller`          `k8s.io/cloud-provider/pkg/controller`         `Yes`          `Partial`

Comment (andrewsykim) [reply keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: Is the `pkg` necessary? Can it just be `k8s.io/cloud-provider/controller`?

File: keps/sig-cloud-provider/0013-removing-in-tree-providers.md [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+| pkg/version                       |                              |                |                  |
+| pkg/volume                        | k8s.io/cloud-provider/volume | Yes            | Partial          |
+| pkg/volume/util                   | k8s.io/cloud-provider/volume | Yes            | Partial          |
+| pkg/volume/util/volumepathhandler | k8s.io/cloud-provider/volume | Yes            | Partial          |
+
+#### Phase 2 - Building CCM from Provider Repos
+
+In Phase 2, providers will be expected to build cloud controller manager from their respective provider repos (`kubernetes/cloud-provider-<provider>`). Provider repos will be frequently synced from their respective staging repos by the kubernetes publishing bot. Development of providers are still done in `kubernetes/kubernetes/staging/cloudprovider/providers`.
+
+#### Phase 3 - Migrating Provider Code to Provider Repos
```
Comment (andrewsykim) [inline keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: Important to note here that we are not proposing that `k8s.io/cloud-provider-<provider>` will be vendored back into `k8s.iokubernetes`

File: keps/sig-cloud-provider/0013-removing-in-tree-providers.md (hunk 2 above)
Comment (stewart-yu) [reply keps/sig-cloud-provider/0013-removing-in-tree-providers.md]: `k8s.io/cloud-provider/controller` ok for me

**Comment by justaugustus** [comment]
REMINDER: KEPs are moving to [k/enhancements](https://github.com/kubernetes/enhancements) on November 30. Please attempt to merge this KEP before then to signal consensus.
For more details on this change, review [this thread](https://groups.google.com/d/msg/kubernetes-dev/LIkZoIqCT20/t515CV2mAgAJ).

Any questions regarding this move should be directed to that thread and not asked on GitHub.

**Comment by andrewsykim** [comment]
Will do my best to update this KEP and have someone review by end of this week. No guarantees though because there's a lot of changes here to be reviewed. If it's not merged by deadline, feel free to close and I'll re-open in the new repo. Thanks @justaugustus!

**Comment by justaugustus** [comment]
KEPs [have moved](https://github.com/kubernetes/enhancements/pull/638) to [k/enhancements](https://github.com/kubernetes/enhancements). 
This PR will be closed and any additional changes to this KEP should be submitted to k/enhancements.
For more details on this change, review [this thread](https://groups.google.com/d/msg/kubernetes-dev/LIkZoIqCT20/t515CV2mAgAJ).

Any questions regarding this move should be directed to that thread and not asked on GitHub.
/close

**Pull request closed without merging**